Display tokens and staking information at dashboard

enjojoy opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

- Display tokens and staking information for each asset.
- Some minor refactors

## Related Issue

Resolve https://github.com/trezor/trezor-suite/issues/14456

## Screenshots:

### Before:


- Piggy has a bellybutton
<img width="519" alt="image" src="https://github.com/user-attachments/assets/7484b7a3-06a1-43a7-a9a3-6c6f25158710">



- Dashboard table
<img width="772" alt="image" src="https://github.com/user-attachments/assets/489b2eec-5039-441e-97cd-75f5d5cc0b53">



- Dashboard grid
<img width="772" alt="image" src="https://github.com/user-attachments/assets/5f0b1253-c19a-4dda-9371-3183e02f6fbd">

### After:

- Piggy doesn't have a bellybutton anymore
![image](https://github.com/user-attachments/assets/a77edd0a-5ed5-4519-b5d2-08ab61308ede)


- Dashboard table
<img width="772" alt="image" src="https://github.com/user-attachments/assets/dabe7db1-666b-4498-b4f4-9abf4ad4b00e">
<img width="1059" alt="image" src="https://github.com/user-attachments/assets/1f6504e8-2299-4bd4-bed4-5c3f03744928">


- Dashboard grid
<img width="1059" alt="image" src="https://github.com/user-attachments/assets/f10a9719-08ac-4481-97cc-8851865baff3">
<img width="805" alt="image" src="https://github.com/user-attachments/assets/9d341d29-3624-4d3b-9fab-0bcb2b1e4632">

tomasklim reviewed (changes requested):
Please validate it against designs provided in Figma

e.g.

Missing "tokens" and "staking" text next to pig and token icons
<img width="968" alt="Screenshot 2024-10-28 at 10 57 04" src="https://github.com/user-attachments/assets/d1bcfe72-778b-4731-a850-afd2fa7b4976">

Figma:
<img width="646" alt="Screenshot 2024-10-28 at 10 57 34" src="https://github.com/user-attachments/assets/8919bd03-8ae2-47ea-bcd6-04fcb6410262">

enjojoy commented:
Thank you, I also fixed the overall table alignment so it's more like in figma. It's in fixup https://github.com/trezor/trezor-suite/pull/14996/commits/b4628e27cc66918ccfacaafee490abb7e52e97b7
<img width="1062" alt="image" src="https://github.com/user-attachments/assets/695f6705-edc7-478b-aff4-5bb6ce6f4e31">

@tomasklim Should I also change Balance to Amount like in the design?

seibei-iguchi commented on packages/suite/src/views/dashboard/AssetsView/AssetCoinLogo.tsx:
```diff
@@ -35,7 +26,7 @@ export const AssetCoinLogo = ({ symbol, assetsFiatBalances, index }: AssetCoinLo
 
     return (
-        <LogoWrapper>
+        <Row alignItems="center" justifyContent="center">
```
alignItems="center" is the default

enjojoy replied:
https://github.com/trezor/trezor-suite/pull/14996/commits/d1b9c30e115df36aba8ad13e537786aa33226844

seibei-iguchi commented on packages/suite/src/views/dashboard/AssetsView/AssetCard/AssetCard.tsx:
```diff
@@ -21,11 +21,25 @@ import { FiatHeader } from 'src/components/wallet/FiatHeader';
+import { FiatCurrencyCode } from '@suite-common/suite-config';
 
 // eslint-disable-next-line local-rules/no-override-ds-component
 const StyledCard = styled(Card)`
```
Please remove this override and the custom padding, use paddingType instead.

seibei-iguchi commented on packages/suite/src/views/dashboard/AssetsView/AssetCard/AssetCard.tsx:
```diff
@@ -34,7 +48,12 @@ const StyledCard = styled(Card)`
 `;
 
 const Content = styled.div`
```
Can be done with Column. The padding is not needed, I think.

enjojoy replied:
https://github.com/trezor/trezor-suite/pull/14996/commits/e57c6a41d63a7760b80706c1b73811aeca389776

seibei-iguchi commented on packages/suite/src/views/dashboard/AssetsView/AssetTable/AssetRow.tsx:
```diff
@@ -49,64 +95,113 @@ export const AssetRow = memo(
+
+                    <Table.Cell>{!isTestnet(symbol) && <TrendTicker symbol={symbol} />}</Table.Cell>
+                    <Table.Cell align="right" colSpan={2}>
+                        <Row gap={16}>
```
Use spacings

enjojoy replied:
https://github.com/trezor/trezor-suite/pull/14996/commits/4458ddcc3a6e480a9f577c96610327de1345259a

seibei-iguchi commented:
I'm not sure this is a good approach, it seems too fragile and magical, using the padding and positioning. I'd rather not use separate rows for this use case and do it inside one row, I think.

seibei-iguchi commented on packages/suite/src/views/dashboard/AssetsView/AssetTable/AssetTable.tsx:
```diff
@@ -1,60 +1,63 @@
+        <Table isRowHighlightedOnHover margin={{ top: spacings.md, bottom: spacings.md }}>
+            <Table.Header>
+                <Table.Row>
+                    <Table.Cell colSpan={1} padding={{ right: spacingsPx.xxs }}></Table.Cell>
```
Why the empty cell?

enjojoy commented:
> I'm not sure this is a good approach, it seems too fragile and magical, using the padding and positioning. I'd rather not use separate rows for this use case and do it inside one row, I think.

What I like about this approach is that it comes with all the paddings and hovers that a row has and I don't have to introduce too much custom stuff. 

![image](https://github.com/user-attachments/assets/3ee51160-566d-4cd9-bf95-01b8c32fbc03)

Regarding custom padding, I'm not sure there's a better way to make the first and second column be as close to each other as in the design. If I'd put them in one column, I think it would be hard to align the name of the coin with "Token" and "Staking" labels, so it looks like they're in the same column, because the token icon is different size than staking and tokens icons. 

That being said, imo it would be ok to remake the rows and put them in one row, even though I don't really see how it's fragile, but I wouldn't change how the columns are done. 

What do you think?

seibei-iguchi commented:
> > I'm not sure this is a good approach, it seems too fragile and magical, using the padding and positioning. I'd rather not use separate rows for this use case and do it inside one row, I think.
> 
> What I like about this approach is that it comes with all the paddings and hovers that a row has and I don't have to introduce too much custom stuff.
> 
> ![image](https://private-user-images.githubusercontent.com/101045932/381045250-3ee51160-566d-4cd9-bf95-01b8c32fbc03.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxOTYyMDEsIm5iZiI6MTczMDE5NTkwMSwicGF0aCI6Ii8xMDEwNDU5MzIvMzgxMDQ1MjUwLTNlZTUxMTYwLTU2NmQtNGNkOS1iZjk1LTAxYjhjMzJmYmMwMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyOVQwOTU4MjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xODYxM2ZkMzNiMTU3NmZhZWVkNmZlNjFjMDU0YjJlMGU5YWRiYTE5YmZlYzY1NjFkZjlhODZlNDM3NjBhNmVmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.I-x9enYcK64-Y3j0sIeGTI4bN3SLU4UIMy-XAWyxx0I)
> 
> Regarding custom padding, I'm not sure there's a better way to make the first and second column be as close to each other as in the design. If I'd put them in one column, I think it would be hard to align the name of the coin with "Token" and "Staking" labels, so it looks like they're in the same column, because the token icon is different size than staking and tokens icons.
> 
> That being said, imo it would be ok to remake the rows and put them in one row, even though I don't really see how it's fragile, but I wouldn't change how the columns are done.
> 
> What do you think?

I'll try to quickly come up with something in code and do a commit, we can then discuss over that:)

seibei-iguchi commented:
@enjojoy So I've made some changes. I've kept the separate rows, but I've cleaned it up a little and made it a little bit more robust and less magical. I'd suggest you abstract the Section component into a separate file, because the only variable part are the top and bottom properties. You can control them via a prop with values like top|middle|bottom so that it renders differently depending on where it is used (tokens, staking etc). Let me know what you think!

seibei-iguchi commented on packages/components/src/components/Table/TableRow.tsx:
```diff
@@ -79,6 +85,7 @@ export const TableRow = ({
     onClick,
     onHover,
     isHighlightedOnHover,
+    isBorderTop = true,
```
perhaps change to hasBorderTop?

enjojoy replied:
https://github.com/trezor/trezor-suite/pull/14996/commits/9f9ea451a2e3b1eadccbb362382b4508d8632267

seibei-iguchi commented:
By the way, wouldn't it be nice if the tokens and stakings row would also link to the respective sections?

enjojoy commented:
> @enjojoy So I've made some changes. I've kept the separate rows, but I've cleaned it up a little and made it a little bit more robust and less magical. I'd suggest you abstract the Section component into a separate file, because the only variable part are the top and bottom properties. You can control them via a prop with values like top|middle|bottom so that it renders differently depending on where it is used (tokens, staking etc). Let me know what you think!

So when there's only one additional row, it's okay, althogh I fixed it a bit (top: calc(50% - 1px);) so the spacings between the end and the beginning of the dotted lines are consistent and it looks like one straight line. 

But when there's two rows, it looks bad either way :( 
Too little (your solution)
![image](https://github.com/user-attachments/assets/2f43323d-8190-45ab-abdf-105f0de9d20b)


Too much (added 1px difference )
![image](https://github.com/user-attachments/assets/5836656d-e624-4613-b6d3-2871412c2ab7)

I think it looked good before, but you didn't like the hardcoded values right?

enjojoy commented:
> By the way, wouldn't it be nice if the tokens and stakings row would also link to the respective sections?

I also thought about it and I think it's a good idea. @tomasklim  What do you think?

enjojoy commented:
> > @enjojoy So I've made some changes. I've kept the separate rows, but I've cleaned it up a little and made it a little bit more robust and less magical. I'd suggest you abstract the Section component into a separate file, because the only variable part are the top and bottom properties. You can control them via a prop with values like top|middle|bottom so that it renders differently depending on where it is used (tokens, staking etc). Let me know what you think!
> 
> So when there's only one additional row, it's okay, althogh I fixed it a bit (top: calc(50% - 1px);) so the spacings between the end and the beginning of the dotted lines are consistent and it looks like one straight line.
> 
> But when there's two rows, it looks bad either way :( Too little (your solution) ![image](https://private-user-images.githubusercontent.com/101045932/381143889-2f43323d-8190-45ab-abdf-105f0de9d20b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAyMDk4NDgsIm5iZiI6MTczMDIwOTU0OCwicGF0aCI6Ii8xMDEwNDU5MzIvMzgxMTQzODg5LTJmNDMzMjNkLTgxOTAtNDVhYi1hYmRmLTEwNWYwZGU5ZDIwYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyOVQxMzQ1NDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04OTI4YTI2YjQyYzdiMTQ5YzNmNjViNTM4NDExZDQxYjE2NTYzZTU3MjVmMjdkMGU0MGFlOWEwZWFkZWIyODMxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.VpA0t_PomnjcvCYqswbWdir5G0mIMSuEJkLZGO5WR9U)
> 
> Too much (added 1px difference ) ![image](https://private-user-images.githubusercontent.com/101045932/381143719-5836656d-e624-4613-b6d3-2871412c2ab7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAyMDk4NDgsIm5iZiI6MTczMDIwOTU0OCwicGF0aCI6Ii8xMDEwNDU5MzIvMzgxMTQzNzE5LTU4MzY2NTZkLWU2MjQtNDYxMy1iNmQzLTI4NzE0MTJjMmFiNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyOVQxMzQ1NDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMzkyOTg1Nzg0M2NhNTZhN2M3MGFjM2Q2NWZiNTJmMTZiNTU1MWUzNTNlYTAzZDY0NjIyZDY0NTI0NjBmMjk5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.0Cnk0c6LaHt7FW1CPHckgXIUPLRVT_KoI8RZC7Fuzuk)
> 
> I think it looked good before, but you didn't like the hardcoded values right?

Nevermind, before it was like this too. I think the only way to fix is is to make a line with absolute position or something
![image](https://github.com/user-attachments/assets/e7cbfb11-1eec-4c91-b14e-cd2d4a24265a)

enjojoy commented:
> > By the way, wouldn't it be nice if the tokens and stakings row would also link to the respective sections?
> 
> I also thought about it and I think it's a good idea. @tomasklim What do you think?

I think it would be a little tricky because when I click on the asset, it takes me to the first asset, but if staking can be on another asset, so we'd have to iterate over it and the same with tokens. So maybe open an issue for it to do it later. Because now these rows don't get the account that has the staking or the tokens

seibei-iguchi commented on packages/suite/src/views/dashboard/AssetsView/AssetTable/AssetStakingRow.tsx:
```diff
@@ -0,0 +1,64 @@
+import { FiatValue } from 'src/components/suite';
+import { Text } from '@trezor/components';
+
+const Section = styled.div<{ $renderBothRows: boolean }>`
```
Can you please abstract this into a separate file and use it in all three types of rows?

enjojoy replied:
https://github.com/trezor/trezor-suite/pull/14996/commits/2ebf646952b19aed134aa8613c51dd5d924c5fb1 
Thanks for a good idea

tomasklim commented on packages/suite/src/views/dashboard/AssetsView/AssetTable/AssetRow.tsx:
```diff
@@ -10,25 +10,56 @@ import {
+        localCurrency,
+        currentFiatRates,
+        accounts,
+    }: AssetTableRowProps & {
+        localCurrency: FiatCurrencyCode;
+        currentFiatRates?: RatesByKey;
+    }) => {
```
I dont like how you reuse `AssetTableRowProps` :( it makes the code really confusing

enjojoy replied:
Understandable, I got rid of this
https://github.com/trezor/trezor-suite/pull/14996/commits/db24b70b62881dce04ca81e408a9e95506a2cc3a

tomasklim commented on packages/suite/src/views/dashboard/AssetsView/AssetTable/AssetTokenRow.tsx:
```diff
@@ -0,0 +1,35 @@
+    network,
+    tokensDisplayFiatBalance,
+}: {
+    tokenIconSetWrapper: React.ReactNode;
+    network: AssetTableRowProps['network'];
+    tokensDisplayFiatBalance: string;
```
missing interface

tomasklim commented on suite-common/wallet-core/src/transactions/transactionsReducer.ts:
```diff
@@ -409,6 +409,15 @@ export const selectAccountHasStaked = (state: TransactionsRootState, account: Ac
+    state: TransactionsRootState,
+    accounts: Account[],
+) => {
+    const accountsThatStaked = accounts.filter(account => selectAccountHasStaked(state, account));
+
+    return accountsThatStaked;
+};
```
```suggestion
    return accounts.filter(account => selectAccountHasStaked(state, account));
};
```

tomasklim commented on packages/suite/src/views/dashboard/AssetsView/AssetCard/AssetCard.tsx:
```diff
@@ -122,21 +132,52 @@ export const AssetCard = ({
 
+    const coinDefinitions = useSelector(state => selectCoinDefinitions(state, symbol));
     const { fiatAmount } = useFiatFromCryptoValue({ amount: cryptoValue, symbol });
+    const stakingAccountsForAsset = stakingAccounts.filter(account => account.symbol === symbol);
+    const accountsThatStaked = useSelector(state =>
+        selectAssetAccountsThatStaked(state, stakingAccountsForAsset),
+    );
+    const assetStakingBalance = accountsThatStaked.reduce((total, account) => {
+        return total.plus(getAccountTotalStakingBalance(account));
+    }, new BigNumber(0));
+    const tokens = getTokens(assetTokens ?? [], symbol, coinDefinitions);
+    const tokensWithRates = enhanceTokensWithRates(
+        tokens.shownWithBalance ?? [],
+        localCurrency,
+        symbol,
+        currentFiatRates,
+    );
+    const sortedTokens = tokensWithRates.sort(sortTokensWithRates);
+    const tokensFiatBalance = sortedTokens.reduce((total, token) => {
+        return total.plus(token?.fiatValue ?? 0);
+    }, new BigNumber(0));
+    const shouldRenderStakingRow = accountsThatStaked.length > 0 && assetStakingBalance.gt(0);
+    const shouldRenderTokenRow = tokens.shownWithBalance?.length > 0 && tokensFiatBalance.gt(0);
```
this code looks like a duplicate of code in `AssetRow.tsx`

Can `AssetCard` and `AssetRow` have a parent which would resolve this logic? and it will just accept some param like `displayType: 'card' | 'row'`

tomasklim replied:
If yes, please do it in separate commit

enjojoy replied:
I feel like it doesn't make sense to create a parent component because card is an independent component and row is a part of a table, but I agree that it makes sense to take out the data handlings because they essentially do the same thing, so I extracted them in a separate function. 
https://github.com/trezor/trezor-suite/pull/14996/commits/59a5790a04f98d8d432a98d107ee7ce1b62fa2d5

tomasklim commented on suite-common/wallet-utils/src/accountUtils.ts:
```diff
@@ -669,6 +669,26 @@ export const getTokensFiatBalance = (
+) => {
+    let totalBalance = new BigNumber(0);
+
+    accounts.forEach(account => {
```
would be nicer to use reduce

```
    accounts.reduce((total, account) => {
        const tokensBalance = getAccountTokensFiatBalance(
            account,
            localCurrency,
            rates,
            account.tokens,
        );
        return total.plus(tokensBalance ?? 0);
    }, new BigNumber(0)).toFixed();
```

if it works

enjojoy replied:
Yes with a slight change, here in a fixup
https://github.com/trezor/trezor-suite/pull/14996/commits/b2b5096161bd3f365cb8e613cc7f173d8a16d47a

tomasklim commented on packages/suite/src/views/dashboard/AssetsView/AssetTable/AssetStakingRow.tsx:
```diff
@@ -0,0 +1,46 @@
+interface AssetStakingRowProps {
+    stakingTotalBalance: string;
+    symbol: NetworkSymbol;
+    renderBothRows: boolean;
```
`shouldRenderBothRows`

tomasklim replied:
or maybe `shouldRenderTokenRow`? as you know that you should render staking at that place

enjojoy replied:
Yes, good catch https://github.com/trezor/trezor-suite/pull/14996/commits/5a47ab41b461e17bd4f7eeafdfd91ebfc2884ae7

tomasklim commented on packages/suite/src/components/wallet/TokenIconSetWrapper.tsx:
```diff
@@ -3,30 +3,57 @@ import {
+                acc[contract] = {
+                    ...token,
+                    balance: balance ?? '0',
+                    fiatValue: fiatValue ?? { s: 0, e: 0, c: [0] },
```
what is s, e, c?

enjojoy replied:
It is a BigNumber as an object, but probably forgot to change it to a normal BigNumber, here it is fixed https://github.com/trezor/trezor-suite/pull/14996/commits/1a6ffd52c4948eb5c17e79078aaab4088ffadd44

tomasklim commented on packages/product-components/src/components/TokenIconSet/TokenIconSet.tsx:
```diff
@@ -49,9 +52,9 @@ export const TokenIconSet = ({ network, tokens }: TokenIconSetProps) => {
         <IconContainer $length={length}>
             {length > 3 && <TokenIconPlaceholder $elevation={elevation} />}
-            {visibleTokens.map(token => (
+            {visibleTokens.map((token, i) => (
```
not needed as duplicated cannot be there right?

enjojoy replied:
Yes, not anymore https://github.com/trezor/trezor-suite/pull/14996/commits/78aaf22d419fdf3f33b44af2ded70d95baf30f2d
